Regression in enum resolving in pyrflx

When trying to run the validation tool on SPDM I hit the following error:
```
------------------------------ RecordFlux Bug ------------------------------
RecordFlux 0.6.0-pre
RecordFlux-parser 0.11.0
z3-solver 4.9.1.0
pydotplus 2.0.2
attrs 21.4.0
pydantic 1.9.1
icontract 2.6.1
ruamel.yaml 0.17.21

Optimized: False

Command: /tmp/tmp.NEqskQfbh4/venv/bin/rflx --no-verification --max-errors=1 validate -o build/validate_static_response_feature_challenge_auth.log -v tests/data/spdm/Response/valid_feature_challenge_auth build/specs/CHALLENGE_AUTH=True,KEY_EXCHANGE=True,RESPOND_IF_READY=True/spdm.rflx SPDM::Response

Traceback (most recent call last):
  File "/tmp/tmp.NEqskQfbh4/venv/lib/python3.9/site-packages/rflx/cli.py", line 202, in main
    args.func(args)
  File "/tmp/tmp.NEqskQfbh4/venv/lib/python3.9/site-packages/rflx/cli.py", line 359, in validate
    Validator(
  File "/tmp/tmp.NEqskQfbh4/venv/lib/python3.9/site-packages/rflx/validator.py", line 99, in validate
    validation_result = self._validate_message(
  File "/tmp/tmp.NEqskQfbh4/venv/lib/python3.9/site-packages/rflx/validator.py", line 283, in _validate_message
    parsed_message.parse(original_message)
  File "/tmp/tmp.NEqskQfbh4/venv/lib/python3.9/site-packages/rflx/pyrflx/typevalue.py", line 882, in parse
    ) = set_field_with_size(current_field_name, current_field_size)
  File "/tmp/tmp.NEqskQfbh4/venv/lib/python3.9/site-packages/rflx/pyrflx/typevalue.py", line 853, in set_field_with_size
    self._set_parsed_value(
  File "/tmp/tmp.NEqskQfbh4/venv/lib/python3.9/site-packages/rflx/pyrflx/typevalue.py", line 1012, in _set_parsed_value
    self._set_checked(field_name, value, message_size)
  File "/tmp/tmp.NEqskQfbh4/venv/lib/python3.9/site-packages/rflx/pyrflx/typevalue.py", line 985, in _set_checked
    check_outgoing_condition_satisfied()
  File "/tmp/tmp.NEqskQfbh4/venv/lib/python3.9/site-packages/rflx/pyrflx/typevalue.py", line 929, in check_outgoing_condition_satisfied
    assert (
AssertionError: unresolved field conditions in Response.Challenge_Auth_Response_Nonce_Data: 2 /= SPDM::Meas_Unsupported
and 1 /= SPDM::No_Measurement, 2 = SPDM::Meas_Unsupported
or 1 = SPDM::No_Measurement

----------------------------------------------------------------------------
```
I bisected the git and could trace the bug back to https://github.com/Componolit/RecordFlux/commit/d049f5f6cef784ad0c83403c4ee033e2e18d3a71.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -115,6 +115,7 @@ PyRFLX:
 
 - Error caused by relations between sequences, opaque fields or aggregates (#964)
 - Undefined attribute in `MessageValue.Field` (#1045)
+- Missing type check for arguments of parameterized message (#1104)
 
 ## [0.5.0] - 2021-08-11
 

--- a/rflx/pyrflx/typevalue.py
+++ b/rflx/pyrflx/typevalue.py
@@ -30,6 +30,7 @@ from rflx.expression import (
 )
 from rflx.identifier import ID
 from rflx.model import (
+    BOOLEAN,
     FINAL,
     INITIAL,
     ByteOrder,
@@ -602,6 +603,23 @@ class MessageValue(TypeValue):
         self._refinements = [*(self._refinements or []), refinement]
 
     def add_parameters(self, parameters: ty.Mapping[str, ty.Union[bool, int, str]]) -> None:
+        expected = set(p.name for p in self._type.parameter_types.keys())
+        added = set(p for p in parameters)
+
+        if expected - added:
+            message = ", ".join(expected - added)
+            raise PyRFLXError(f"missing parameter values: {message}")
+
+        if added - expected:
+            message = ", ".join(added - expected)
+            raise PyRFLXError(f"unexpected parameter values: {message}")
+
+        def check_type(valid: bool) -> None:
+            if not valid:
+                raise PyRFLXError(
+                    f'message argument for "{name}" has invalid type "{type(value).__name__}"'
+                )
+
         enum_literals = {
             str(l): str(t.package * l)
             for t in self._type.types.values()
@@ -612,29 +630,27 @@ class MessageValue(TypeValue):
         expr: Expr
         for name, value in parameters.items():
             if isinstance(value, bool):
+                check_type(self.model.parameter_types[Field(name)] == BOOLEAN)
                 expr = TRUE if value else FALSE
             elif isinstance(value, int):
+                check_type(isinstance(self.model.parameter_types[Field(name)], Integer))
                 expr = Number(value)
             elif isinstance(value, str):
+                check_type(
+                    isinstance(self.model.parameter_types[Field(name)], Enumeration)
+                    and self.model.parameter_types[Field(name)] != BOOLEAN
+                )
                 if value in enum_literals:
                     expr = Literal(enum_literals[value])
                 else:
                     assert value in enum_literals.values()
                     expr = Literal(value)
             else:
-                raise PyRFLXError(f"{type(value)} is no supported parameter type")
+                raise PyRFLXError(
+                    f'message argument for "{name}" has unsupported type "{type(value).__name__}"'
+                )
+
             params[Variable(name)] = expr
-
-        expected = set(p.name for p in self._type.parameter_types.keys())
-        added = set(p.name for p in params if isinstance(p, Variable))
-
-        if expected - added:
-            message = ", ".join(expected - added)
-            raise PyRFLXError(f"missing parameter values: {message}")
-
-        if added - expected:
-            message = ", ".join(added - expected)
-            raise PyRFLXError(f"unexpected parameter values: {message}")
 
         self._parameters = params
         if not self._skip_verification:

--- a/tests/unit/pyrflx_test.py
+++ b/tests/unit/pyrflx_test.py
@@ -1426,13 +1426,46 @@ def test_parameterized_message_no_verification() -> None:
     assert message_unv.bytestring == bytes(8)
 
 
-def test_parameterized_message_invalid_type(parameterized_package: Package) -> None:
+def test_parameterized_message_unsupported_type(parameterized_package: Package) -> None:
     with pytest.raises(
-        PyRFLXError, match=f"^pyrflx: error: {type(bytes())} is no supported parameter type$"
+        PyRFLXError,
+        match='^pyrflx: error: message argument for "Length" has unsupported type "bytes"$',
     ):
         parameterized_package.new_message(
             "Message",
-            {"Length": bytes(8)},  # type: ignore[dict-item]
+            {
+                "Length": bytes(8),  # type: ignore[dict-item]
+                "Tag_Mode": "Without_Tag",
+                "Tag_Value": "Parameterized::Tag_A",
+                "Use_Tag": True,
+            },
+        )
+
+
+def test_parameterized_message_invalid_type(parameterized_package: Package) -> None:
+    with pytest.raises(
+        PyRFLXError, match='^pyrflx: error: message argument for "Tag_Mode" has invalid type "int"$'
+    ):
+        parameterized_package.new_message(
+            "Message",
+            {
+                "Length": 8,
+                "Tag_Mode": 2,
+                "Tag_Value": "Parameterized::Tag_A",
+                "Use_Tag": True,
+            },
+        )
+    with pytest.raises(
+        PyRFLXError, match='^pyrflx: error: message argument for "Use_Tag" has invalid type "str"$'
+    ):
+        parameterized_package.new_message(
+            "Message",
+            {
+                "Length": 8,
+                "Tag_Mode": "Without_Tag",
+                "Tag_Value": "Parameterized::Tag_A",
+                "Use_Tag": "True",
+            },
         )
 
 

